src/adapters/curve/products/pool/curvePoolAdapter.ts

```diff
@@ -133,15 +171,46 @@ export class CurvePoolAdapter
     return protocolToken
   }
 
-  /**
-   * Update me.
-   * Add logic that finds the underlying token rates for 1 protocol token
-   */
   protected async getUnderlyingTokenConversionRate(
-    _protocolTokenMetadata: Erc20Metadata,
-    _blockNumber?: number | undefined,
+    protocolTokenMetadata: Erc20Metadata,
+    blockNumber: number | undefined,
   ): Promise<UnderlyingTokenRate[]> {
-    throw new NotImplementedError()
+    const { poolAddress, underlyingTokens, protocolToken } =
+      await this.fetchPoolMetadata(protocolTokenMetadata.address)
+
+    const metaRegistryContract = MetaRegistry__factory.connect(
+      CURVE_META_REGISTRY_CONTRACT,
+      this.provider,
+    )
+
+    const balances = await metaRegistryContract['get_balances(address)'](
+      poolAddress,
+      { blockTag: blockNumber },
+    )
+
+    const lpTokenContract = Erc20__factory.connect(
+      protocolToken.address,
+      this.provider,
+    )
+
+    const supply = await lpTokenContract.totalSupply({ blockTag: blockNumber })
+
+    return balances.slice(0, underlyingTokens.length).map((balance, index) => {
+      const underlyingToken = underlyingTokens[index]!
+
+      const formattedBalance = formatUnits(balance, underlyingToken.decimals)
+      const formattedSupply = formatUnits(supply, protocolToken.decimals)
+
+      const price = +formattedBalance / +formattedSupply
+
+      return {
+        type: TokenType.Underlying,
+        underlyingRateRaw: BigInt(
+          Math.round(price * 10 ** underlyingToken.decimals),
+        ),
+        ...underlyingToken,
+      }
```

Comment:
Ok...

So the balance and the supply are being converted first to "numbers" (really to strings, then to numbers, but that makes no difference) and then multiplied again by the decimals and converted to `bigint`. Very dodgy.

To calculate `formatBalance`, what's essentially doing is `balance / 10 ** underlyingToken.decimals`, and then at the end we are multiplying by `10 ** underlyingToken.decimals`, which I think means we could skip the first division.

What I suggest we do instead:

```suggestion
      const formattedBalance = formatUnits(balance, underlyingToken.decimals)
      const formattedSupply = formatUnits(supply, protocolToken.decimals)

      const underlyingRateRaw = balance / (supply / 10 ** protocolToken.decimals)

      return {
        type: TokenType.Underlying,
        underlyingRateRaw,
        ...underlyingToken,
      }
```